fix(chart): lock digest differs when dependency build with Helm 2 and then Helm 3

hickeyma opened the pull request:
When apiVersion v1 chart dependencies are built with Helm 2 and then built with Helm 3, the lock digests differ. To avoid this issue, a dependency update is forced.

Fixes #7233

hickeyma committed:
Fix issue with apiVersion v1 lock digest

When apiVersion v1 chart dependencies are built with Helm 2
and then built with Helm 3, the lock digests differ. To avoid
this issue, a depdendency update is forced.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

karuppiah7890 commented:
Do you think it's possible to could add a failing test here for the issue #7233 ?

TBBle commented on pkg/downloader/manager.go:
```diff
@@ -86,6 +86,11 @@ func (m *Manager) Build() error {
+		// If lock digest differs and chart is apiVersion v1, it maybe because the lock was built
+		// with Helm 2 and should therefore be regenerated
+		if c.Metadata.APIVersion == chart.APIVersionV1 {
+			return m.Update()
```
I feel like Update is the wrong thing here, because that might pull in *different* semver-matching versions of packages, ignoring the versions *locked* in the Lockfile.

Can we detect an APIVersionV1 Hash and recognise it?

technosophos replied:
I agree that update is a "destructive" operation, and thus should not be used here. 

i am wondering if we have to add a function on the Manager that re-calculates the hashes for v2. But even there, I am a little bit uncertain.

TBBle replied:
How hard would it be to make `resolver.HasReq(req, nil)` generate the same hash as Helm v2? Did `chart.Dependencies` change a lot?

mattfarina replied:
The semver package is a major version change between Helm v2 and v3. Behind semver some rules were mildly modified in how they calculate versions. In order to calculate the rules the same as Helm v2 we would need to pull in semver v1 and perform the calculation using that. You can read about the changes [here](https://github.com/Masterminds/semver/blob/master/CHANGELOG.md#300-2019-09-12).

hickeyma replied:
@mattfarina Thanks for the details about the semver.

@technosophos @TBBle FYI, I went with the "nuclear" approach first as I could be certain about the result. I am not however happy with this approach (or even doing a hashing function to map from v2 to v3) as I feel we are circumventing a valid error being raised.

I think that @technosophos suggestion in https://github.com/helm/helm/issues/7233#issuecomment-579023372 about adding a warning with a specific message might be the best approach.

I am open to ideas on how best to work this.

TBBle replied:
@mattfarina Those changes seem to be about how the semver is matched? The Helm v2 HashReq was only looking at the pre-match expression (as a string), so I don't think it's a major issue if we accept whatever was selected at `helm dep update` time, if the has tells us that the requirements.yaml (and for v3 requirements.lock) are the same as at that time, even if `helm dep up` *now* would pick different versions.

The structure itself (`Dependency`) is unchanged, unless hashing picks up the package of the type.

I don't think this is a valid failure to raise as an error, personally. The chart *is* valid under Helm v2, a newer version of Helm shouldn't suddenly consider the chart *invalid*. We actually have the converse problem that `APIVersion1` charts produced with Helm v3 would fail this check with Helm v2, except I think the check didn't exist in Helm v2?

hickeyma replied:
> The chart is valid under Helm v2, a newer version of Helm shouldn't suddenly consider the chart invalid.

@TBBle I appreciate your justification on why this is needed.

I have pushed changes which will now check against v2 style hash. let me know what you think.

hickeyma committed:
Check against Helm v2 hash

Handle scenario where dependency hash was generated by Helm v2
but need to do a dependency build with Helm v3.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Updated and ready for review again.

hickeyma committed:
Add unit test

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

mattfarina reviewed (commented):
@hickeyma A large part of this is rather large test charts. Do we want these full charts in the codebase or would something simpler suffice?

hickeyma committed:
Refactor unit test

Refactor unit test to use an existing chart as dependency

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Good point @mattfarina, thanks. Its only after the commit, I realised that it had mushroomed! :)

I have refactored it now and it is ready for review again.

TBBle commented on pkg/downloader/manager.go:
```diff
@@ -86,7 +87,17 @@ func (m *Manager) Build() error {
+				return errors.New("the lock file is out of sync with the dependencies file. Please update the dependencies")
+			}
+		} else {
+			return errors.New("the lock file is out of sync with the dependencies file. Please update the dependencies")
```
The error messages could be more specific, and name the relevant files. We know whether we're an APIVersionV1 or APIVersionV2 chart at this point.

hickeyma replied:
Good point. Updated the messages to include the filenames.

TBBle commented on cmd/helm/dependency_build_test.go:
```diff
@@ -100,3 +100,22 @@ func TestDependencyBuildCmd(t *testing.T) {
+	}
+
+	// Make sure the child chart file got downloaded.
+	expect := filepath.Join(chartName, "charts/alpine-0.1.0.tgz")
```
The chart's committed. So if the download doesn't happen, this still passes? Or does this logic rely on `helm dep build` removing the existing file first?

Or is alpine-0.1.0.tgz supposed to be committed outside this Chart, so that `helm dep build` copies it in?

hickeyma replied:
Hmm, good point. Removing the check as it doesn't make lot of sense as including the tarball.

hickeyma committed:
Update after review

Comments:
- https://github.com/helm/helm/pull/7261#discussion_r373827088
- https://github.com/helm/helm/pull/7261#discussion_r373827250

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks for the reviews @mattfarina @TBBle.

@technosophos Are you ok with this to merge?

technosophos reviewed (approved):
I think this is ready to merge. Thanks for rewriting that!

hickeyma commented:
Thanks @technosophos for reviews.
